<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html>
<head>
   <meta http-equiv="Content-Type" content="text/html; charset=iso-8859-1">
   <meta name="description" content="GAUL Documentation: A Simple Example">
   <meta name="keywords" content="genetic algorithm, GA, tutorial, examples, documentation, GAUL">
   <meta name="author" content="Stewart Adcock">
   <meta name="copyright" content="© Stewart Adcock">
   <meta name="generator" content="vi">
   <meta http-equiv="Content-Language" content="en">
   <title>GAUL: Genetic Algorithm Utility Library</title>
</head>
<body text="#000000" bgcolor="#D8D8FF" link="#222288" vlink="#222288" alink="#9999FF">

<a NAME="top"></a>

<center>
<script language="JavaScript">
<!--
google_ad_client = 'pub-0728071010451673';
google_ad_width = 468;
google_ad_height = 60;
google_ad_format = '468x60_as';
google_ad_channel ="8508682197";
// -->
</script>
<script language="JavaScript" src="http://pagead2.googlesyndication.com/pagead/show_ads.js">
</script>
</center>

<table BORDER=0 CELLSPACING=0 CELLPADDING=0 WIDTH="100%">
  <tr>
    <td ALIGN=CENTER BGCOLOR="#000000"><font face="Arial,Helvetica" color="#FFFFFF"> [ <a href="../index.php"><font face="Arial,Helvetica" color="#FFFFFF"><b>Home</b></font></a> | <a href="../documentation.html"><font face="Arial,Helvetica" color="#FFFFFF"><b>Docs</b></font></a> | <a href="../news.html"><font face="Arial,Helvetica" color="#FFFFFF"><b>News</b></font></a> | <a href="../resources.html"><font face="Arial,Helvetica" color="#FFFFFF"><b>Resources</b></font></a> | <a href="../downloads.html"><font face="Arial,Helvetica" color="#FFFFFF"><b>Download</b></font></a>] </font></td>
  </tr>
</table>
<table BORDER=0 CELLSPACING=0 CELLPADDING=0 WIDTH="100%">
  <tr>
    <td ALIGN=CENTER BGCOLOR="#2222AA"><b><font face="Arial,Helvetica" color="#FFFFFF">GAUL Documentation: A Simple Example</font></b></td>
  </tr>
</table>

<p><a NAME="intro"></a></p>
<table BORDER=0 CELLSPACING=0 CELLPADDING=2 WIDTH="100%">
<tr>
<td>
<p><b>Preface</b></p>
<p>This is the start of the GAUL tutorial.  It will explain and demonstrate how to use GAUL to develop your own programs with genetic or evolutionary algorithms.</p>
<p>For the purpose of this tutorial, it is assumed that the reader has a reasonable understanding of the C programming language.  When alternative language wrappers for the GAUL library are made available, additional versions of this tutorial will be written to cover those.</p>
<p>It is also assumed that the reader already understands the basic principles of genetic algorithms.  Genetic algorithms are briefly described on <a href="../intro.html">this page</a>.</p>
<p>During this tutorial several GA-based programs are developed to perform particular tasks.  These programs are intended to demonstrate various features of the the GAUL library, so they are rarely optimal solutions to their tasks.  Consequently, you are encouraged to tweak the examples and experiment to improve on the solutions and to see how the various parameters interact.  You may wish to refer to the accompaning <a href="../gaul_reference_guide.html">reference guide</a> while working through this tutorial.</p>

<p><b>A minimal program using GAUL</b></p>
<p>Below is a guide to producing a really simple GA-based program.  This example doesn't make use of any of GAUL's more advanced features, but is intended to demonstrate the ease with which a minimal application may be created.</p>
The aim of this example is to generate the final sentence from Chapter 3 of Darwin's &quot;The Origin of Species&quot;, entitled &quot;Struggle for Existence&quot;.  Taking the &quot;Monkeys with typewriters&quot; style approach would basically never solve this problem.</p>
<p>For reference, slightly extended source code for this example is distributed in the gaul-examples package as source-code.  It may be found as <a href="http://cvs.sourceforge.net/cgi-bin/viewcvs.cgi/*checkout*/gaul/gaul-examples/src/struggle.c?only_with_tag=HEAD&amp;view=markup">struggle.c</a>.  The only difference between these codes is that the distributed example runs the GA procedure fifty times instead of just once.</p>

<p><b>Annotated source code</b></p>
<p>The public header file for the GAUL library is &quot;gaul.h&quot; and this must be included by any code that will use GAUL:</p>
<pre>
#include &lt;gaul.h&gt;
</pre>
<p>This public header file defines all of the GAUL specific functions and datatypes.  The two most important datatypes, &quot;entity&quot; and &quot;population&quot;, contain all the data required to handle the individuals in a population and that population itself.  In GAUL, each population has a set of GA parameters, such as mutation rates.  These datastructures are described further below.  The term &quot;entity&quot; is synonymous with &quot;individual&quot; or &quot;potential solution&quot;.  These terms will be used interchangably during this tutorial.</p>
<p>In this example, we define the target solution in a global string:</p>
<pre>
char *target_text="When we reflect on this struggle, we
 may console ourselves with the full belief, that the war
 of nature is not incessant, that no fear is felt, that
 death is generally prompt, and that the vigorous, the
 healthy, and the happy survive and multiply.";
</pre>
<p>The first thing that is needed is the <i>objective function</i>, the code for determining the &quot;fitness&quot; of each potential solution.  Objective functions are used to define the fitness landscape on which the GA search will occur.  The term &quot;objective function&quot; is synonymous with the term &quot;fitness function&quot;.  For this example program the most appropriate form of chromosome is a single array of chars (a string).  The objective function simply has to compare each character in an individual's chromosome with the corresponding character in the target string to derive the fitness score:<p>
<pre>
boolean struggle_score(population *pop, entity *entity)
  {
  int           k;              /* Loop variable over all alleles. */

  entity->fitness = 0.0;

  /* Loop over alleles in chromosome. */
  for (k = 0; k < pop->len_chromosomes; k++)
    {
    if ( ((char *)entity->chromosome[0])[k] == target_text[k])
      entity->fitness+=1.0;
    /*
     * Component to smooth function, which helps a lot in this case:
     * Comment it out if you like.
     */
    entity->fitness += (127.0-fabs(((char *)entity->chromosome[0])[k]
                                -target_text[k]))/50.0;
    }

  return TRUE;
  }
</pre>
<p>The objective function must return &quot;TRUE&quot; upon successful evaluation of the fitness, and &quot;FALSE&quot; otherwise.  Cases for the later would include the attempted scoring of an invalid chromosome.  This example function assumes that the chromosome is always valid.  You will notice that the chromosome must be cast to the correct type - this is because the core routines in GAUL are totally independant of the actual datatypes.  You will notice that GAUL actually uses an array of chromosomes, thus enabling efficient support for multiple chromosome descriptions of any problem.</p>
<p>The &quot;entity&quot; datastructure is used to store details about individual solutions.</p>
<p>The calculated fitness score is placed into the &quot;fitness&quot; field of the &quot;entity&quot; structure.</p>
<p>The &quot;population&quot; datastructure is used to store the global information about the solutions, including the GA parameters and so on.<p>
<p>The &quot;population&quot; structure contains several useful fields.  Some particular fields that are likely to be of use in most applications are &quot;len_chromosomes&quot; (the user-defined length of each chromosome), &quot;num_chromosomes&quot; (the user-defined number of chromosomes per individual), and &quot;size&quot; (the total number of individuals in the population).</p>
<p>GAUL uses its own random number generator which must be initialised, either by calling &quot;random_init()&quot; or by calling &quot;random_seed()&quot;.  This is generally done at the start of any application using GAUL.  We also need to define at least one population structure.  In this example, we perform both of those tasks in the code's main function.</p>
<pre>
int main(int argc, char **argv)
  {
  population *pop=NULL;	/* The population of solutions. */

  random_seed(2003);	/* Random seed requires any integer parameter. */
</pre>
<p>Now we need to define the population and specify the prefered variant of genetic algorithm.  GAUL provides a selection of high-level convenience wrappers to generate populations with default parameters.  These are named &quot;ga_genesis_XXX()&quot;, where XXX is the chromosome datatype.  In this example, we will use &quot;ga_genesis_char()&quot;, because we are using the character-array chromosome type.</p>
<p>The majority of parameters to these genesis functions are pointers to callback functions which are used to perform various portions of the GA algorithm.  In this way it is trivial to use you own code to, say, mutate the chromosome using a new, or problem specific, routine.  In this example, however, only GAUL's built-in functions are utilised.</p>
<p>It is perfectly valid to specific a different set of callback functions for each population that you define.  This makes GAUL usable in applications that perform multiple, different, search or optimisation tasks.  The parameters for the genesis functions will be described and illustrated throughout this tutorial.</p>
<pre>
    pop = ga_genesis_char(
       250,                      /* const int              population_size */
       1,                        /* const int              num_chromo */
       strlen(target_text),      /* const int              len_chromo */
       NULL,                     /* GAgeneration_hook      generation_hook */
       NULL,                     /* GAiteration_hook       iteration_hook */
       NULL,                     /* GAdata_destructor      data_destructor */
       NULL,                     /* GAdata_ref_incrementor data_ref_incrementor */
       struggle_score,           /* GAevaluate             evaluate */
       ga_seed_printable_random, /* GAseed                 seed */
       NULL,                     /* GAadapt                adapt */
       ga_select_one_sus,        /* GAselect_one           select_one */
       ga_select_two_sus,        /* GAselect_two           select_two */
       ga_mutate_printable_singlepoint_drift, /* GAmutate  mutate */
       ga_crossover_char_allele_mixing, /* GAcrossover     crossover */
       NULL,                     /* GAreplace              replace */
       NULL                      /* void *                 userdata */
            );

    ga_population_set_parameters(
       pop,                     /* population              *pop */
       GA_SCHEME_DARWIN,        /* const ga_class_type     class */
       GA_ELITISM_PARENTS_DIE,  /* const ga_elitism_type   elitism */
       0.9,                     /* double                  crossover */
       0.2,                     /* double                  mutation */
       0.0                      /* double                  migration */
                              );
</pre>
<p>The second and third parameters to &quot;ga_population_set_parameters()&quot; specify the type of GA variant to apply.  These parameters are described, in some detail, later in this tutorial.  The specific parameters used here cause GAUL to use a standard Darwinian GA with no elitism.</p>
<p>The final three parameters to &quot;ga_population_set_parameters()&quot; specify the crossover, mutation and migration ratios, respectively.  The migration ratio is ignored in the GA simulation performed here, but is required for island-model GAs.  The exact meaning of these ratios depends upon the particular operator functions, from above, that are being used.
<p>This call to &quot;ga_population_set_parameters()&quot; could even be skipped, if you are happy with the default GA parameters defined within GAUL.</p>
<p>Now we can invoke the actual GA evolution routines:</p>
<pre>
    ga_evolution(
       pop,                     /* population              *pop */
       500                      /* const int               max_generations */
              );
</pre>
<p>The function &quot;ga_evolution()&quot; performs a standard generation-based GA.  An alternative is &quot;ga_evolution_steady_state()&quot; which performs a standard steady-state GA.  This example is modified for use with this steady-state GA in <a href="http://cvs.sourceforge.net/cgi-bin/viewcvs.cgi/*checkout*/gaul/gaul-examples/src/struggle_ss.c?only_with_tag=HEAD&amp;view=markup">struggle_ss.c</a>.</p>
<p>And finally, we deallocate the population structure, and are finished:<p>
<pre>
  printf( "The final solution found was:\n", i);
  printf( "%s\n", ga_chromosome_char_to_staticstring(pop, ga_get_entity_from_rank(pop,0)));
  printf( "Fitness score = %f\n", ga_get_entity_from_rank(pop,0)->fitness);

  ga_extinction(pop);	/* Deallocates all memory associated with
			 * the population and it's entities.
			 */

  exit(EXIT_SUCCESS);
  }
</pre>
<p>The &quot;ga_get_entity_from_rank()&quot; function may be used to extract particular solutions from the population structure.  The ranks run from 0 to <b>N</b>-1, inclusive, where <b>N</b> is the current population size.</p>
<p>The return value of &quot;ga_evolution()&quot; is the number of generations actually performed.
</td>
</tr>
</table>

<p>[ <a href="#top">BACK TO TOP</a> | <a href="evolution.html">NEXT SECTION OF TUTORIAL</a> ]</p>

<p><a NAME="footer"></a></p>
<table BORDER=0 CELLSPACING=0 CELLPADDING=2 WIDTH="100%">
  <tr>
    <td ALIGN=CENTER BGCOLOR="#000000"><font face="Arial,Helvetica" color="#FFFFFF"> [ <a href="../index.php"><font face="Arial,Helvetica" color="#FFFFFF"><b>Home</b></font></a> | <a href="../documentation.html"><font face="Arial,Helvetica" color="#FFFFFF"><b>Docs</b></font></a> | <a href="../news.html"><font face="Arial,Helvetica" color="#FFFFFF"><b>News</b></font></a> | <a href="../resources.html"><font face="Arial,Helvetica" color="#FFFFFF"><b>Resources</b></font></a> | <a href="../downloads.html"><font face="Arial,Helvetica" color="#FFFFFF"><b>Download</b></font></a>] </font></td>
  </tr>
</table>

<!--   Footer, Links etc.  -->
<table BORDER=0 CELLSPACING=2 CELLPADDING=5 WIDTH="100%">
  <tr>
    <td width="20%" ALIGN=CENTER></td>
    <td width="20%" ALIGN=CENTER>Hosted by:<br><A href="http://sourceforge.net"><IMG src="http://sourceforge.net/sflogo.php?group_id=25766" width="88" height="31" border="0" alt="SourceForge Logo"></A></td>
    <td width="20%" ALIGN=CENTER><A href="http://gaul.sourceforge.net/"><IMG src="../images/gaul_button.png" width="88" height="31" border="0" alt="GAUL Button"></A></td>
    <td width="20%" ALIGN=CENTER><a href="http://validator.w3.org/check/referer"><img border="0" src="../images/valid-html401.png" alt="Valid HTML 4.01!" width="88" height="31"></a></td>
    <td width="20%" ALIGN=CENTER></td>
  </tr>
</table>

<table BORDER=0 CELLSPACING=0 CELLPADDING=0 WIDTH="100%">
  <tr><td><hr noshade width="100%"></td></tr>
  <tr>
    <td ALIGN=CENTER><font face="Arial,Helvetica" size=-1>&copy; 2001-2005, &quot;Stewart Adcock" &lt;<a href="MAILTO:stewart_DOESNOTLIKESPAM@linux-domain.com">stewart@linux-domain.com</a>&gt; All rights reserved.</font></td>
  </tr>
  <tr>
    <td ALIGN=CENTER><font face="Arial,Helvetica" size=-1>All trademarks on this page are probably owned by their respective companies.  You can insert you own favourite disclaimer here.</font></td>
  </tr>
</table>

</body>
</html>

